src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,176 @@
+
+                if (documentSnapshot is null ||
+                    !_documentVersionCache.TryGetDocumentVersion(documentSnapshot, out documentVersion))
+                {
+                    documentVersion = null;
+                }
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = null,
+                    HostDocumentVersion = null
+                };
+            }
+
+            var unmappedDiagnostics = request.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d)).ToArray();
+            if (!filteredDiagnostics.Any())
+            {
+                // No diagnostics left after filtering.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = Array.Empty<Diagnostic>(),
+                    HostDocumentVersion = documentVersion
+                };
+            }
+
+            return await MapDiagnosticsAsync(
+                request,
+                filteredDiagnostics,
+                documentVersion,
+                documentSnapshot).ConfigureAwait(false);
+
+            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
+            static bool CanDiagnosticBeFiltered(Diagnostic d) =>
+                (DiagnosticsToIgnore.Contains(d.Code?.String) &&
+                 d.Severity != DiagnosticSeverity.Error);
+        }
+
+        private async Task<RazorDiagnosticsResponse> MapDiagnosticsAsync(
+            RazorDiagnosticsParams request,
+            Diagnostic[] filteredDiagnostics,
+            int? documentVersion,
+            DocumentSnapshot documentSnapshot)
+        {
+            if (request.Kind != RazorLanguageKind.CSharp)
+            {
+                // All other non-C# requests map directly to where they are in the document.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = filteredDiagnostics,
+                    HostDocumentVersion = documentVersion,
+                };
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            var mappedDiagnostics = new List<Diagnostic>();
+
+            for (var i = 0; i < filteredDiagnostics.Length; i++)
+            {
+                var diagnostic = filteredDiagnostics.ElementAt(i);
+                var projectedRange = diagnostic.Range;
+
+                if (codeDocument.IsUnsupported() ||
+                    !_documentMappingService.TryMapFromProjectedDocumentRange(
+                        codeDocument, projectedRange,
```

Comment:
nit: I'd break these into separate lines, feels weird for these to be side-by-side when everything else isn't

Reply:
Yeah totally, no idea why I kept these together. Nice catch!